clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java

```diff
@@ -0,0 +1,226 @@
+            if (t instanceof RuntimeException)
+                this.exception.set(Optional.of((RuntimeException) t));
+            else
+                this.exception.set(Optional.of(new RuntimeException(t)));
+        } finally {
+            close();
+            log.debug("{} closed", getClass());
+        }
+    }
+
+    /**
+     * Process event from a single poll
+     */
+    void runOnce() {
+        this.inflightEvent = maybePollEvent();
+        log.debug("processing application event: {}", this.inflightEvent);
+        if (this.inflightEvent.isPresent() && maybeConsumeInflightEvent(this.inflightEvent.get())) {
+            // clear inflight event upon successful consumption
+            this.inflightEvent = Optional.empty();
+        }
+
+        // if there are pending events to process, poll then continue without
+        // blocking.
+        if (!applicationEventQueue.isEmpty() || inflightEvent.isPresent()) {
+            networkClient.poll(time.timer(0));
+            return;
+        }
+        // if there are no events to process, poll until timeout. The timeout
+        // will be the minimum of the requestTimeoutMs, nextHeartBeatMs, and
+        // nextMetadataUpdate. See NetworkClient.poll impl.
+        networkClient.poll(time.timer(timeToNextHeartbeatMs(time.milliseconds())));
+    }
+
+    private long timeToNextHeartbeatMs(long nowMs) {
+        // TODO: implemented when heartbeat is added to the impl
+        return 100;
+    }
+
+    private Optional<ApplicationEvent> maybePollEvent() {
+        if (this.inflightEvent.isPresent() || this.applicationEventQueue.isEmpty()) {
+            return this.inflightEvent;
+        }
+        return Optional.ofNullable(this.applicationEventQueue.poll());
+    }
+
+    /**
+     * ApplicationEvent are consumed here.
+     * @param event an {@link ApplicationEvent}
+     * @return true when successfully consumed the event.
+     */
+    private boolean maybeConsumeInflightEvent(ApplicationEvent event) {
+        log.debug("try consuming event: {}", Optional.ofNullable(event));
+        switch (event.type) {
+            case NOOP:
+                process((NoopApplicationEvent) event);
+                return true;
+            default:
+                inflightEvent = Optional.empty();
+                log.warn("unsupported event type: {}", event.type);
+                return true;
+        }
+    }
+
+    /**
+     * Processes {@link NoopApplicationEvent} and equeue a
+     * {@link NoopBackgroundEvent}. This is intentionally left here for
+     * demonstration purpose.
+     * @param event a {@link NoopApplicationEvent}
+     */
+    private void process(NoopApplicationEvent event) {
+        backgroundEventQueue.add(new NoopBackgroundEvent(event.message));
+    }
+
+    public boolean isRunning() {
+        return this.running;
+    }
+
+    public void wakeup() {
+        networkClient.wakeup();
+    }
+
+    public void close() {
+        this.wakeup();
```

Comment:
Does the order of these two lines (`wakeup` and `running = false`) want to be swapped?

`close` is called by the polling thread, right? If `wakeup` is called, it will throw an exception, but then the background thread might potentially loop back around to `runOnce` before the call to set `running` to `false` by the polling thread.